### PR TITLE
image compare test suite

### PR DESCRIPTION
**Pull request by kevinb7**
@lmccart this is far from finished.  I wanted to do an early pull request so you could see where I'm headed with this.  I've ported roughly 20% of the image compare tests from processing-js.  I've put them in a separate folder while I work on them.

I'm using your ```same``` calculations from ```renderTest```.  Each test can be configured to have a different acceptable ```sameness``` (now that I think of it, it's really the "difference" that I'm calculating... I'll change that).  There are a couple of variables at the top ```showAll``` and ```showFailures``` which can be used to show the actual image beside the reference image.  Here's a screen shot of how that looks, __note__ currently there 50 tests passing and about 15 being skipped:

screen shot 2015-02-09 at 6 27 39 pm

I noticed that github doesn't show the contents of .pde files so I think I'll switch the extension to .js.

### CODE REVIEW

File: image_compare/test_spec.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+      { file: 'blendcolor-difference.pde', tolerance: 0.015, skip: true },
+      { file: 'blendcolor-dodge.pde', tolerance: 0.015, skip: true },
+      { file: 'blendcolor-exclusion.pde', tolerance: 0.015, skip: true },
+      { file: 'blendcolor-hard_light.pde', tolerance: 0.015, skip: true },
+      { file: 'blendcolor-lightest.pde', tolerance: 0.015, skip: true },
+      { file: 'blendcolor-multiply.pde', tolerance: 0.015, skip: true },
+      { file: 'blendcolor-overlay.pde', tolerance: 0.015, skip: true },
+      { file: 'blendcolor-screen.pde', tolerance: 0.015, skip: true },
+      { file: 'blendcolor-soft_light.pde', tolerance: 0.015, skip: true },
+      { file: 'blendcolor-subtract.pde', tolerance: 0.015, skip: true }
```
Comment (kevinb7) [inline image_compare/test_spec.js]: Do we want to implement [blendColor](http://processingjs.org/reference/blendColor_/)?

File: image_compare/test_spec.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    ]},
+    
+    { file: 'blur.pde', tolerance: 0.015 },
+    { file: 'bounce.pde', tolerance: 0.015, timeout: 1000 },
+    // uses "class" keyword
+    { file: 'bouncy-bubbles.pde', tolerance: 0.015, skip: true },
+    { file: 'brightness.pde', tolerance: 0.015 },
+    { file: 'brightness-2.pde', tolerance: 0.015 },
+    // TODO: randomSeed algorithm differs from processing-js
+    { file: 'brownian.pde', tolerance: 0.015 }
```
Comment (kevinb7) [inline image_compare/test_spec.js]: I ran this test a couple of times and it outputs the same result each time which is what ```randomSeed``` should be doing, it's just the random numbers are different from processing-js (which I believe re-implements the same algorithm as the original Processing).

**Comment by kevinb7** [comment]
@lmccart I know you were concerned about differences between browsers.  One way to address that is to make sure that we always use the same browser to run the tests.  I've had some experience using docker images with Firefox to do this.  I set up the docker image with ubuntu, Firefox, and FreeType to match the versions on travis-ci.  This way users can run the image compare tests locally regardless of what OS they're running on.  

This will allow use to do exact pixel comparisons which are more reliable.  Before we can do this though we need a way to generate our own set of reference images because we have antialiasing on by default which doesn't match the current reference images which don't.

I'd like to get the whole suite ported first before making these enhancements to the test harness because even though it's not exact it can still pick up issues.

File: image_compare/test_spec.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+  
+  function getRefData(code) {
+    var line = code.split('\n')[0];
+    var regex = /\/\/\[[0-9]+\,[0-9]+\]/;
+    var matches = regex.exec(line);
+
+    var len = matches[0].length;
+    return line.substring(len).split(',').map(function (val) {
+      return parseInt(val);
+    });
```
Comment (kevinb7) [inline image_compare/test_spec.js]: The first line of each .pde file contains a comment that looks something like this:
```//[100,100]204,204,204,255,204,204,204,255,...```
The first line contains the reference image data.

**Comment by kevinb7** [comment]
The current test suite takes about 3 seconds in the browser (2012 MacBook Air, Core i5 @ 1.7 GHz).  In general I've found the browser to be faster than phantomjs.

**Comment by lmccart** [comment]
hey @kevinb7, thanks for this, it's looking great, and so sorry for the delay. one question, are the ones in the ref/ folder ported already or unported? I think it'd be best to remove all references to processingjs and processing (extension .pde, size() calls, etc) and have a clean set of images we can run against the ported p5.js code.

**Comment by kevinb7** [comment]
@lmccart no worries, I've been busy with other stuff and haven't had time to finish these off.  I'll put some time into it this week.  All of the ones in the ```ref``` folder have been ported already except for maybe one or two which needed a non-trivial amount of additional work.  There's a whole bunch that I haven't copied over from __processingjs__ yet.  I'm going for the low hanging fruit first.  Once that's complete I'll move on to generating a clean set of images.

**Comment by lmccart** [comment]
great!

**Comment by lmccart** [comment]
I think this was a work in progress and not ready to be merged, is that correct? I'm going to close for now, please reopen if I'm mistaken or it becomes ready for merge.

**Comment by kevinb7** [comment]
@lmccart sorry I don't have time to finish this off.

**Pull request closed without merging**